src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+      Abs decls' ans' <- locally $ occNest a decls ans
+      ty' <- occ accessOnce ty  -- TODO Is free variable counting enough here?
+      countFreeVarsAsOccurrences effs
+      return $ Block (BlockAnn ty' effs) decls' ans'
+
+-- TODO(optimization) Could reuse the free variable caching from dce here too.
+
+data ElimResult n where
+  ElimSuccess :: Abs (Nest Decl) Atom n -> ElimResult n
+  ElimFailure :: Decl n l -> UsageInfo -> Abs (Nest Decl) Atom l -> ElimResult n
+
+occNest :: Access n -> Nest Decl n l -> Atom l
+        -> OCCM n (Abs (Nest Decl) Atom n)
+occNest a decls ans = case decls of
+  Empty -> Abs Empty <$> occ a ans
+  Nest d@(Let _ binding) ds -> do
+    isPureDecl <- isPure binding
+    -- Note that we only ever analyze the abs below under this refreshAbs,
+    -- which will remove any references to b upon exit (it happens
+    -- because refreshAbs of StateT1 triggers hoistState, which we
+    -- implement by deleting the entries that can't hoist).
+    dceAttempt <- refreshAbs (Abs d (Abs ds ans))
+      \d'@(Let b' (DeclBinding _ _ expr')) (Abs ds' ans') -> do
+        exprIx <- summaryExpr $ sink expr'
+        extend b' exprIx do
+          below <- locally $ occNest (sink a) ds' ans'
+          accessInfo <- getAccessInfo $ binderName d'
+          let usage = usageInfo accessInfo
+          let dceAttempt = case isPureDecl of
+               False -> ElimFailure d' usage below
+               True  ->
+                 -- Or hoistUsingCachedFVs in the monad, if we decide to do
+                 -- that optimization
+                 case hoist d' below of
+                   HoistSuccess below' -> ElimSuccess below'
+                   HoistFailure _ -> ElimFailure d' usage below
+          return dceAttempt
+    case dceAttempt of
+      ElimSuccess below' -> return below'
+      ElimFailure (Let b' binding') usage (Abs ds'' ans'') -> do
+        -- Using accessOnce here, instead of the computed Access for
+        -- the decl's binder.  This means that variable bindings cut
+        -- occurrence analysis, and each binding is considered for
+        -- inlining separately.
+        DeclBinding _ ty expr <- occ accessOnce binding'
+        let binding'' = DeclBinding (OccInfo usage) ty expr
+        return $ Abs (Nest (Let b' binding'') ds'') ans''
+
+instance HasOCC DeclBinding where
+  occ a (DeclBinding ann ty expr) = do
+    expr' <- occ a expr
+    -- TODO(optimization) Is free variable counting enough for types?
+    ty' <- occ accessOnce ty
+    return $ DeclBinding ann ty' expr'
+
+instance HasOCC Expr where
+  occ a expr = case expr of
+    (TabApp array (NE.toList -> ixs)) -> do
+      (a', ixs') <- go a ixs
+      array' <- occ a' array
+      return $ TabApp array' $ NE.fromList ixs'
+    (Case scrut alts ty effs) -> do
+      scrut' <- occ accessOnce scrut
+      scrutIx <- summary scrut
+      (alts', innerFVs) <- unzip <$> mapM (isolated . occAlt a scrutIx) alts
```

Comment:
I am a little surprised that you pass in the same IxExpr for the scrutinee to every case, and that ends up being bound as the IxExpr for the case binder. The case binder is definitely not that same as the scrutinee, since it has the top constructor eliminated.

Reply:
It doesn't end up being bound as the IxExpr for the case binder: the case binder is an `unknown` applied to the scrutinee.  This is a conservative approximation, due to not having constructors in `IxExpr` for eliminating the top constructor; and I guess that if I could statically eliminate it then the whole case would have been removed by case-of-known-constructor.

Added a comment to this effect.

Reply:
In that case I think I don't understand this code. Why are you passing `scrutIx` to `occAlt` then?

Reply:
Future-proofing for when there will be a constructor for `IxExpr` that represents the elimination that a case expression does.  And even barring that, if we get `unknown` to detect effects and emit `Deterministic` sometimes, then the case binder will also be tagged `Deterministic` rather than `IxAll`, which may help precision at some point.

Reply:
Right, but so you do assume that `scrutIx` always will be either `Deterministic` or `IxAll`, right? Perhaps worth adding that as a comment, with an explanation that proves why this will be the case?

Reply:
No, I don't assume anything about the `scrutIx` that's in scope inside `instance HasOCC SExpr`.  But until we have a representation in `IxExpr` for "case analysis removed constructor foo", the `scrutIx` in scope in `occAlt` will end up being `Deterministic <something>` or `IxAll`, based on how well `unknown` works.